libsplinter/src/network/ref_map.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::{error, fmt};
+
+/// A map that will keep track of the number of times an id has been added, and only remove the
+/// id once the reference count is 0.
+///
+/// The RefMap also keeps track of redirects. This allows ids to be changed while keeping the same
+/// refefrence count.
+pub struct RefMap {
+    // id to reference count
+    references: HashMap<String, u64>,
+    redirects: HashMap<String, String>,
+}
+
+impl RefMap {
+    pub fn new() -> Self {
+        RefMap {
+            references: HashMap::new(),
+            redirects: HashMap::new(),
+        }
+    }
+
+    pub fn add_ref(&mut self, id: String) -> u64 {
+        // check if id is for a current id or a redirect
+        let ref_id = {
+            if let Some(ref_id) = self.redirects.get(&id) {
+                ref_id.clone()
+            } else {
+                // if the id is not in the reference or redirects, it is a new reference
```

Comment:
These comments are slightly out-of-date since we're not checking for an existing reference here.

Reply:
removed this comment, as the above comment covers it